engine/consensus/matching/engine.go

```diff
@@ -263,17 +272,57 @@ func (e *Engine) process(originID flow.Identifier, event interface{}) error {
 
 // onReceipt processes a new execution receipt.
 func (e *Engine) onReceipt(originID flow.Identifier, receipt *flow.ExecutionReceipt) error {
+	// when receiving a receipt, we might not be able to verify it if its previous result
+	// is unknown.  In this case, instead of dropping it, we store it in the pending receipts
+	// mempool, and process it later when its parent result has been received and processed.
+	// therefore, if a receipt is processed, we will check if it is the previous results of
+	// some pending receipts, if there are, then processing them one after another, as onReceipt
+	// is behind a lock
+	receiptID := receipt.ID()
+	resultID := receipt.ExecutionResult.ID()
+
+	processed, err := e.onCurrentReceipt(receipt)
+	if err != nil {
+		e.log.Error().
+			Err(err).
+			Hex("receipt_id", receiptID[:]).
+			Hex("result_id", resultID[:]).
+			Hex("origin_id", originID[:]).
+			Msg("could not process receipt")
+		return nil
+	}
+
+	if !processed {
+		return nil
+	}
+
+	childReceipts := e.pendingReceipts.ByPreviousResultID(resultID)
+	e.pendingReceipts.Rem(receiptID)
+
+	for _, childReceipt := range childReceipts {
+		// recursively processing the child receipts, since onReceipt
+		// is logging error internal already, we could ignore the returned
+		// error here
+		_ = e.onReceipt(childReceipt.ExecutorID, childReceipt)
+	}
+
+	return nil
+}
+
+func (e *Engine) onCurrentReceipt(receipt *flow.ExecutionReceipt) (bool, error) {
```

Comment:
I would prefer returning error that can be a sentinel error in case it wasn't processed, something like `DuplicatedEntryError` and `OutdatedInputError`

Reply:
We can discuss the error handling in 387. The reason I added a `bool` is to use `error` as exception. When there is no exception, then I only care about the `bool`. 

So that when we receipt a duplicated receipt or an outdated receipt, it won't be logged as "error", since they are not error. The caller of `onCurrentReceipt` will only log error if the error is indeed an exception. If there is no exception, `onCurrentReceipt` itself will log the receipt has been stored.